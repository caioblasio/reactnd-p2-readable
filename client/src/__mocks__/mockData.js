const testPosts = {

  "8xf0y6ziyjabvozdd253nd": {
    id: '8xf0y6ziyjabvozdd253nd',
    timestamp: 1467166872634,
    title: 'Udacity is the best place to learn React',
    body: 'Everyone says so after all.',
    author: 'thingtwo',
    category: 'categoryone',
    voteScore: 6,
    deleted: false,
    commentCount: 2
  },
  "6ni6ok3ym7mf1p33lnez": {
    id: '6ni6ok3ym7mf1p33lnez',
    timestamp: 1468479767190,
    title: 'Learn Redux in 10 minutes!',
    body: 'Just kidding. It takes more than 10 minutes to learn technology.',
    author: 'thingone',
    category: 'categorytwo',
    voteScore: -5,
    deleted: false,
    commentCount: 0
  }
}

const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]

const testComments = [
  {
    id: '894tuq4ut84ut8v4t8wun89g',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1468166872634,
    body: 'Hi there! I am a COMMENT.',
    author: 'thingtwo',
    voteScore: 6,
    deleted: false,
    parentDeleted: false
  },
  {
    id: '8tu4bsun805n8un48ve89',
    parentId: "8xf0y6ziyjabvozdd253nd",
    timestamp: 1469479767190,
    body: 'Comments. Are. Cool.',
    author: 'thingone',
    voteScore: -5,
    deleted: false,
    parentDeleted: false
  }
  
]

/* JSON headers for API requests */
const jsonHeaders = new Headers({
	Accept: 'application/json',
	'Content-Type': 'application/json'
});

export { testPosts, testCategories, testComments, jsonHeaders };