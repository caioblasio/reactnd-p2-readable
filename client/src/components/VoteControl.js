import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { addVoteToPost } from '../actions/posts';
import { addVoteToComment } from '../actions/comments';

import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import ThumbUp from '@material-ui/icons/ThumbUp';
import ThumbDown from '@material-ui/icons/ThumbDown';

import { checkValue, saveValue } from '../utils/locaStorage';

class VoteControl extends Component {

  state = {
    currentVote: checkValue('votes', this.props.id)
  }

  score = this.props.voteScore;

  handleVote = newVote => {
    const { currentVote } = this.state

    if (currentVote && currentVote === newVote) {
      this.cancel(newVote)
    } else if (currentVote) {
      this.cancel(currentVote)
      this.apply(newVote)
    } else {
      this.apply(newVote)
    }
  }

  cancel = vote => {
    if (vote === 'upVote') {
      this.handleDispatchVote('downVote')
    } else if (vote === 'downVote') {
      this.handleDispatchVote('upVote')
    }

    this.setState({ currentVote: null })
    saveValue('votes', this.props.id, null)
  }

  apply = vote => {
    saveValue('votes', this.props.id, vote)
    this.handleDispatchVote(vote)
    this.setState({ currentVote: vote })
  }


  handleDispatchVote = vote => {
    
    const { id, addVoteToPost, addVoteToComment, type } = this.props;
    vote === 'upVote' ? this.score++ : this.score--;
    const changes = { voteScore: this.score }

    type === 'post'
      ? addVoteToPost(id, changes, vote)
      : addVoteToComment(id, changes, vote)
    
  }

  render() {
    const { currentVote } = this.state,
      { voteScore } = this.props;

    return (
      <Fragment>
        <IconButton 
          aria-label="Like"
          onClick={() => this.handleVote('upVote')}
          color={currentVote === 'upVote'? 'secondary' : 'default'}
        >
          <ThumbUp />
        </IconButton>
        <Typography component="p">
          {voteScore}
        </Typography>
        <IconButton 
          aria-label="Dislike"
          onClick={() => this.handleVote('downVote')}
          color={currentVote === 'downVote'? 'secondary' : 'default'}
        >
          <ThumbDown />
        </IconButton>
      </Fragment>
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addVoteToPost: (id, changes, vote) => {
      dispatch(addVoteToPost(id, changes, vote))
    },
    addVoteToComment: (id, changes, vote) => {
      dispatch(addVoteToComment(id, changes, vote))
    },
  }
}

export default connect(null, mapDispatchToProps)(VoteControl);