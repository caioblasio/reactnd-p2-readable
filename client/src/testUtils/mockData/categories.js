export const testCategories = [
  {
    name: 'Category One',
    path: 'categoryone'
  },
  {
    name: 'Category Two',
    path: 'categorytwo'
  }
]