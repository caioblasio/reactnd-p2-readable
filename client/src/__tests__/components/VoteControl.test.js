import React from 'react';
import toJson from 'enzyme-to-json';
import { testPosts, testComments, jsonHeaders } from '../../__mocks__/mockData';
import VoteControl from '../../components/VoteControl';
import IconButton from '@material-ui/core/IconButton';
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store';

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);


describe('<VoteControl />', () => {

  let wrapper, store;


  beforeAll(() => {

    const initialState = {
      posts: testPosts
    }

    store = mockStore(initialState);

  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders the component', () => {
    
    const testPost = testPosts[Object.keys(testPosts)[0]];

    wrapper = shallow(
      <VoteControl 
        type="post"
        voteScore={testPost.voteScore}
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    expect(toJson(component)).toMatchSnapshot();

  });

  it('handles a first vote', () => {
    
    const testPost = testPosts[Object.keys(testPosts)[0]];

    fetch
      .mockResponse(JSON.stringify(testPost), { jsonHeaders })

    wrapper = shallow(
      <VoteControl 
        type="post"
        voteScore={testPost.voteScore}
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    component.find(IconButton).first().simulate('click');
    expect(component.state().currentVote).toEqual('upVote')
    expect(fetch.mock.calls).toHaveLength(1)

  });

  it('handles downVote after an upVote', () => {
    
    const testPost = testPosts[Object.keys(testPosts)[0]];

    fetch
      .mockResponse(JSON.stringify(testPost), { jsonHeaders })

    wrapper = shallow(
      <VoteControl 
        type="post"
        voteScore={testPost.voteScore}
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    component.find(IconButton).last().simulate('click');
    expect(component.state().currentVote).toEqual('downVote')
    expect(fetch.mock.calls).toHaveLength(2)

  });

  it('handles a cancel vote', () => {
    
    const testPost = testPosts[Object.keys(testPosts)[0]];

    fetch
      .mockResponse(JSON.stringify(testPost), { jsonHeaders })

    wrapper = shallow(
      <VoteControl 
        type="post"
        voteScore={testPost.voteScore}
        id={testPost.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    component.find(IconButton).last().simulate('click');
    expect(component.state().currentVote).toEqual(null)
    expect(fetch.mock.calls).toHaveLength(1)

  });


  it('handles vote to comment correctly', () => {
    
    const testComment = testComments[Object.keys(testComments)[0]];

    fetch
      .mockResponse(JSON.stringify(testComment), { jsonHeaders })

    wrapper = shallow(
      <VoteControl 
        type="comment"
        voteScore={testComment.voteScore}
        id={testComment.id}
        store={store}
      />
    );
    const component = wrapper.dive();
    component.find(IconButton).first().simulate('click');
    expect(component.state().currentVote).toEqual('upVote')
    expect(fetch.mock.calls[0][0]).toEqual(`/api/comments/${testComment.id}`);
    expect(fetch.mock.calls).toHaveLength(1)

  });



});