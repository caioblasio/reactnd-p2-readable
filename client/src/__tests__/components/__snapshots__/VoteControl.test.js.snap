// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VoteControl /> renders the component 1`] = `
<Fragment>
  <WithStyles(IconButton)
    aria-label="Like"
    color="default"
    onClick={[Function]}
  >
    <pure(ThumbUpIcon) />
  </WithStyles(IconButton)>
  <WithStyles(Typography)
    component="p"
  >
    6
  </WithStyles(Typography)>
  <WithStyles(IconButton)
    aria-label="Dislike"
    color="default"
    onClick={[Function]}
  >
    <pure(ThumbDownIcon) />
  </WithStyles(IconButton)>
</Fragment>
`;
